syntax = "proto3";

package tensorflow.data.model.proto;
option cc_enable_arenas = true;

message Model {
  // Counter used for generating new node IDs.
  int64 id_counter = 1;
  // Nodes of this model.
  repeated Node node = 2;
  // The ID of the output node.
  int64 output = 3;
};

message Node {
  // The node ID.
  int64 id = 1;
  // The node name.
  string name = 2;
  // Input node IDs.
  repeated int64 input = 3;
  // Output node ID.
  int64 output = 4;
  // Number of elements produced by the node.
  int64 num_elements = 5;
  // The CPU time spent by running threads of this node.
  int64 processing_time = 6;
  // Key-value store for node metadata (e.g. batch size or parallelism).
  map<string, int32> metadata = 7;
};
